### PR TITLE
Add ability to schedule recurring incremental backups

### LINKED ISSUE #303: Add scheduling support for backups
Currently, each ‘solrbackup’ resource triggers a one-time backup of the specified collections.  This is great for many usecases that only need backups on an ad-hoc basis, but it falls short on SaaS or other hosted usecases that often take backups on a regular schedule as a basic precaution.

Operator users should be able to configure a schedule (either in the existing ‘solrbackup’ CRD, or in a new ‘solrbackupschedule’ resource type) so that backups can be set up to recur without further intervention.  Ideally some sort of “maxBackupsRetained” setting would also be added to garbage-collect old backups that are no longer needed I guess?

### PR DESCRIPTION
**Pull request by HoustonPutman**
Resolves #303 

Still needs a bit of love, but the bulk of the work should be there.

TODO:

- [x] Documentation & Changelog entry
- [x] Cleaning up logic around restarts
- [x] Rebase after other backup PRs are merged
- [x] Manual testing

### CODE REVIEW

File: api/v1beta1/solrbackup_types.go [hunk 1]
```diff
@@ -64,6 +71,29 @@ func (spec *SolrBackupSpec) withDefaults(backupName string) (changed bool) {
...
+	Schedule string `json:"schedule"`
+
+	// Define the number of backup points to save for this backup at any given time.
+	// The oldest backups will be deleted if too many exist when a backup is taken.
+	// If not provided, this defaults to 10.
+	//
+	// +kubebuilder:default:=10
+	// +kubebuilder:validation:Minimum:=1
+	// +optional
+	MaxSaved int `json:"maxSaved,omitempty"`
```
Comment (gerlowskija) [inline api/v1beta1/solrbackup_types.go:94]: [0] This is mostly subjective preference, but I'd vote for a smaller default value here.  For collections with a lot of churn, 10 backups might be a lot of data to store by default.  Would you be open to a lower default like 3 or 5?

**Review by gerlowskija — CHANGES_REQUESTED** [review]
This looks awesome!

All of the code itself looks good (aside from tests which I imagine you just haven't gotten to yet).  I do have two suggestions/question about the feature overall though: 

1. Confusingly, incremental backup 'deletion' doesn't really do a full delete: instead it removes only the metadata associated with the backup, potentially leaving index-files behind.  Any backed up index-files that are no longer used must then be deleted by a separate garbage-collection process called "[purging](https://solr.apache.org/guide/8_10/collection-management.html#deletebackup-parameters)."
IMO we should handle this purging process for users: either by triggering a "purge" on every backup, or by exposing a `purgeEvery` setting alongside `maxSaved` to allow a less aggressive approach. 
2. I can imagine a lot of users might use recurrence but want some way to temporarily disable it while they do a perf test, upgrade, bulk data loading, etc.  What do you think about putting an "enabled" flag in `BackupRecurrence` to let users temporarily disable things without totally deleting the solrbackup?

File: api/v1beta1/solrbackup_types.go (hunk 1 above)
Comment (HoustonPutman) [reply api/v1beta1/solrbackup_types.go:94]: Very open. I literally just chose a number here. I am very happy to differ to whatever you think it should be. Give the call and it is done.

**Comment by HoustonPutman** [comment]
1. I'm content with a `purgeEvery` option in Solr. I'm not sure I understand why you would want un-garbage collected index files laying about, but that's definitely a different issue. We can definitely go ahead and purge after every backup until the `purgeEvery` option is added. (Oh now I realize you mean a `purgeEvery` option in the CRD. I think we can assume that would be equivalent to `saveEvery`.)
I did talk with @anshumg about a ticket that saw some issues with purging that does make me a bit wary to do it always. https://issues.apache.org/jira/browse/SOLR-15706
We should make sure that's not an issue before going forward with that.
2. That's a good idea. I do think I want to have it "enabled" by default. Which is easy to do with those kubebuilder annotations.

**Comment by gerlowskija** [comment]
> now I realize you mean a `purgeEvery` option in the CRD. I think we can assume that would be equivalent to saveEvery.

Yep, that's what I was suggesting - an option in the CRD to expose the option that's already available in Solr.  That SOLR-15703 you pointed to looks very serious 😬 .  I'll take a look and hopefully get the Solr concern out of the way so we can consider exposure in the operator.

When you mention "saveEvery" I assume you mean maxSaved?  Or is there some other "saveEvery" param I missed?

**Comment by HoustonPutman** [comment]
> When you mention "saveEvery" I assume you mean maxSaved? Or is there some other "saveEvery" param I missed?

I mis-typed. I meant `maxSaved` 🙂

**Comment by HoustonPutman** [comment]
@gerlowskija, given that https://issues.apache.org/jira/browse/SOLR-15706 is not released yet, maybe we just wait to add in the `purge` call until the next Solr Operator release.

File: api/v1beta1/solrbackup_types.go (hunk 1 above)
Comment (HoustonPutman) [reply api/v1beta1/solrbackup_types.go:96]: I set the default to 5

**Comment by HoustonPutman** [comment]
@gerlowskija almost forgot, but I added the `disabled` flag and it works really well!

**Comment by killdash9** [comment]
> @gerlowskija, given that https://issues.apache.org/jira/browse/SOLR-15706 is not released yet, maybe we just wait to add in the `purge` call until the next Solr Operator release.

That issue has been fixed.  Should we add purgeUnused to the solr operator's backup functionality?

**Comment by HoustonPutman** [comment]
Absolutely, you should create a new issue for that @killdash9 !

**Pull request merged**